mcos/covariance_transformer.py

```diff
@@ -0,0 +1,211 @@
+from abc import ABC, abstractmethod
+import numpy as np
+from sklearn.neighbors import KernelDensity
+from scipy.optimize import minimize
+import pandas as pd
+
+
+class AbstractCovarianceTransformer(ABC):
+    """
+    Abstract class for transforming a covariance matrix
+    """
+
+    @abstractmethod
+    def transform(self, cov: np.array, n_observations: int) -> np.array:
+        """
+        Transforms a covariance matrix
+        :param cov: covariance matrix
+        :param n_observations: number of observations used to create the covariance matrix
+        :return: transformed covariance matrix
+        """
+
+        pass
+
+
+class CovarianceMatrixDeNoiser(AbstractCovarianceTransformer):
+
+    def transform(self, cov: np.array, n_observations: int) -> np.array:
+        """
+        De-noises a covariance matrix as outlined in section 4.2 of "A Robust Estimator of the Efficient Frontier"
+        :return: transformed covariance matrix
+        """
+        return self._de_noise_covariance_matrix(cov, n_observations)
+
+    def _de_noise_covariance_matrix(
+            self,
+            covariance_matrix: np.array,
+            n_observations: int,
+            bandwidth: float = .25
```

Comment:
Let's put bandwidth as a constructor parameter.

Reply:
for sure!